### PR TITLE
De-deduplicate request types by extracting request type into separate interface

### PR DESCRIPTION
**Pull request by declanvong**
(attempt 2 😄 )

### CODE REVIEW

File: src/app.ts [hunk 1]
```diff
@@ -240,6 +240,10 @@ class TypescriptTextWriter implements ITypescriptTextWriter {
   public method(name: string, parameters: { parameter: string, type: string | TypescriptWriterCallback }[], returnType: string, singleLine = false) {
     this.writer.startIndentedLine(`${name}(`);
 
+    if (!singleLine) {
```
Comment (declanvong) [inline src/app.ts]: Oops, just branched off the previous one so this is still here. Agree with your comment, let's get automatic prettier set up after 👍

File: src/app.ts [hunk 2]
```diff
@@ -487,74 +496,100 @@ export class App {
...
+              if (requestParameters.resource == null && requestBodyType) {
+                writer.comment('Request body');
+                writer.property('resource', requestBodyType, true);
+              }
+            });
+          });
+
+          const methodReturn = getMethodReturn(method, schemas);
+          if (requestParameters.resource != null && requestBodyType) {
+            // If the request has an existing 'resource' property, we drop the additional request body 'resource' parameter
```
Comment (declanvong) [inline src/app.ts]: Added your duplicate `resource` property fix here. Do you remember which Google API had a non-body `resource` property in the request so I can test it out?

Comment (Maxim-Mazurok) [reply src/app.ts]: Don't remember.. maybe, bigquery or something like that. Anyway, you could just generate all types and run DefinetlyTyped tests, all typings except for DialogFlow should pass. I'll review in more detail tomorrow, it's getting late here :)
Thanks for PR!

**Comment by Maxim-Mazurok** [comment]
I see this error for [containeranalysis:v1alpha1](https://containeranalysis.googleapis.com/$discovery/rest?version=v1alpha1):
```
Error:(1076, 13) TS2687: All declarations of 'name' must have identical modifiers.
```
It's because now we have two interfaces named `ListOccurrencesResourceRequest`.

And that's because there are 
```
resources.projects.resources.notes.resources.occurrences.methods.list.parameters.name
```
and 
```
resources.projects.resources.occurrences.methods.list.parameters.name
```

I think to fix that, we should be more specific in `requestTypeName` generation (include parent names). Probably, google-api-nodejs-client style (`Something$Something$SomeRequest`) would prevent most of name collisions.

Same for [dlp:v2](https://dlp.googleapis.com/$discovery/rest?version=v2), [dfareporting:v3.4](https://www.googleapis.com/discovery/v1/apis/dfareporting/v3.4/rest) and [dataflow:v1b3](https://dataflow.googleapis.com/$discovery/rest?version=v1b3), I believe.

---

In [healthcare:v1beta1](https://healthcare.googleapis.com/$discovery/rest?version=v1beta1) I have error with dash-case:
```
Observation-lastn(request: Observation-lastnFhirResourceRequest): Request<HttpBody>;
```
It should've been fixed in this commit: https://github.com/Maxim-Mazurok/google-api-typings-generator/pull/14/commits/32cc056b86ac44fe4170f1766b431f6f40524b79
Please, check that these fixes are included.

Let me know if I can help with anything, thanks for your efforts!

**Comment by Maxim-Mazurok** [comment]
@declanvong I will try to fix this PR later if you don't mind. Right now I want to focus on #15

**Comment by declanvong** [comment]
sounds good! this is just an optimisation PR to reduce the type bloat, we can always add this later :)

**Comment by Maxim-Mazurok** [comment]
Hello, @declanvong ! Could you, please, try to update this PR with the current master to see if it works? I would really like to see this one merged before we start updating types automatically. Probably it would be easier to open new PR, because a lot has been changed in app.ts

**Comment by declanvong** [comment]
👌 i'll try to take a look this week! super busy though, so don't count on it 😅

**Pull request closed without merging**